### PR TITLE
Apply enacted treasury withdrawals

### LINKED ISSUE #3736: Apply TreasuryWithdrawals
Money transfers from the Treasury to the reward accounts from proposals that have been enacted must happen right after enactment.

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Closes #3736 

Also add lenses `ensWithdrawalsL`, `ensTreasuryL` and `epochStateTreasuryL`.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs [hunk 1]
```diff
@@ -82,8 +81,8 @@ enactmentTransition = do
           wdrlsNoNetworkId = Map.mapKeys getRwdCred wdrls
       pure
         st
-          { ensWithdrawals = Map.unionWith (<>) wdrlsNoNetworkId $ ensWithdrawals st
-          , ensTreasury = ensTreasury st <-> wdrlsAmount
+          { ensWithdrawals = Map.unionWith (+) wdrlsNoNetworkId $ ensWithdrawals st
+          , ensTreasury = ensTreasury st - wdrlsAmount -- TODO: This can be removed as it is an unused field
```
Comment (aniketd) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: Let me know whether you think this is correct, so I will go ahead and remove this field.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 2]
```diff
@@ -187,6 +191,20 @@ updateNumDormantEpochs ls =
...
+applyEnactedWithdrawals epochState enactedState =
+  let enactedWithdrawals = enactedState ^. ensWithdrawalsL
+      totalWithdrawals = fold enactedWithdrawals
+      rewardsUView = RewDepUView $ epochState ^. epochStateUMapL
+      refunds = rewardsUView ◁ enactedWithdrawals
+      unsuccessfulRefunds = enactedWithdrawals `Map.difference` refunds
+      epochStateRewardsApplied =
+        epochState
+          & epochStateUMapL .~ (rewardsUView ∪+ refunds)
+          & epochStateTreasuryL %~ (<> fromCompact (fold unsuccessfulRefunds - totalWithdrawals))
```
Comment (aniketd) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: Note that here `totalWithdrawals >= unsuccessfulRefunds`

**Comment by aniketd** [comment]
This needs a bit of rework (potential underflow and some known guarantees).

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs (hunk 1 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs]: This is wrong indeed.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs (hunk 2 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: Wrong.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 3]
```diff
@@ -187,6 +191,32 @@ updateNumDormantEpochs ls =
...
+      unsuccessfulRefunds = enactedWithdrawals `Map.difference` refunds
+      epochStateRewardsApplied =
+        epochState
+          -- The use of the partial function `compactCoinOrError` is justified here because
+          -- 1. the decoder for coin at the proposal-submission boundary has already confirmed we have a compactible value
+          -- 2. the refunds and unsuccessful refunds together do not exceed the current treasury value
+          & epochStateUMapL .~ (rewardsUView ∪+ (compactCoinOrError <$> refunds))
+          & epochStateTreasuryL %~ (<> fold unsuccessfulRefunds)
+      -- `enactedState` already has `totalWithdrawals` subtracted in ENACT, so
+      -- we only reset the withdrawals here.
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: Both, the comment and implementation are wrong. 
ENACT state does not have access to the treasury

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 4]
```diff
@@ -187,6 +191,32 @@ updateNumDormantEpochs ls =
...
+applyEnactedWithdrawals epochState enactedState =
+  let enactedWithdrawals = enactedState ^. ensWithdrawalsL
+      rewardsUView = RewDepUView $ epochState ^. epochStateUMapL
+      refunds = rewardsUView ◁ enactedWithdrawals
+      unsuccessfulRefunds = enactedWithdrawals `Map.difference` refunds
+      epochStateRewardsApplied =
+        epochState
+          -- The use of the partial function `compactCoinOrError` is justified here because
+          -- 1. the decoder for coin at the proposal-submission boundary has already confirmed we have a compactible value
+          -- 2. the refunds and unsuccessful refunds together do not exceed the current treasury value
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]:
```suggestion
          -- 2. the refunds and unsuccessful refunds together do not exceed the current treasury value
          -- This is enforce by the `ENACT` rule
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 5]
```diff
@@ -187,6 +191,32 @@ updateNumDormantEpochs ls =
         then ls & lsCertStateL . certVStateL . vsNumDormantEpochsL +~ 1
         else ls
 
+-- | Apply TreasuryWithdrawals to the EpochState
+--
+--   acnt' = record acnt { treasury = treasury + UTxOState.fees utxoSt
+--                                  + getCoin unclaimed + donations ∸ totWithdrawals }
+--
+-- We only apply the `unclaimed` (`unsuccessful` in our case) refunds to the treasury here
+-- The subtraction of the `totWithdrawals` is already applied in ENACT
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: Same here. This is wrong
```suggestion
```

**Review by lehins — CHANGES_REQUESTED** [review]
You are still missing some parts from the spec, like this one:
https://github.com/input-output-hk/formal-ledger-specifications/blob/d8260e001e248725a01335253ce33f35d1656fe7/src/Ledger/Chain.lagda#L84

We'll talk more about it during the meeting, but I think you haven't looked deeply into how `ensTreasury` field works in EnactState and how it relates to `epochStateTreasuryL

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 6]
```diff
@@ -288,11 +322,10 @@ epochTransition = do
     removedProps =
       foldl' lookupAction mempty $
         Seq.fromList (Set.toList rsRemoved)
-    lState = returnProposalDeposits removedProps $ esLState es'
+    lState = returnProposalDeposits removedProps $ esLState epochState'
     es'' =
-      es'
-        { esAccountState = acnt'
```
Comment (aniketd) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:294]: `acnt'` is already part of the `EpochState` a few lines above. And after that we update it in `applyEnactedWithdrawals` and so, it need not be updated here again.

**Review by lehins — APPROVED** [review]
Perfect! :+1:

**Pull request merged**